Add Connext SwapClient support

erkarl opened the pull request.

Commit:
feat: add Connext SwapClient

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+      }
+    }
+    if (!currency) {
+      throw new Error(`could not find currency for tokenAddress ${tokenAddress}`);
```
This might make sense to add under `errors.ts`

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+    { rHash: string, units: number, expiry?: number, currency?: string },
+  ) => {
+    if (!expectedCurrency) {
+      throw new Error('cannot add invoice, expected currency is missing');
```
This could be another standard "missing currency" error in errors.ts

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+  public address?: string;
+  /** A map of currency symbols to token addresses. */
+  public tokenAddresses = new Map<string, string>();
+  private expectedIncomingTransfers = new Map<string, ExpectedIncomingTransfer>();
```
It'd be good to have a comment here I think to explain that this is basically the equivalent of invoices.

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+      ) {
+        this.logger.debug(`accepting incoming transfer with rHash: ${rHash}, units: ${units}, timelock ${timelock} and currency ${currency}`);
+        this.emit('htlcAccepted', rHash, units, currency);
+        // TODO: should we also store this in the database?
```
I also have this question. Should we even be storing all expected incoming transfers into the database? If xud crashes while we are expecting transfers, would we need them when we come back online?

erkarl replied:
> Should we even be storing all expected incoming transfers into the database?

I think we can derive the expected incoming transfers from the `SwapDeal` since we're persisting every step of the deal.

> If xud crashes while we are expecting transfers, would we need them when we come back online?

Definitely.

erkarl replied:
After thinking about it for a while... should the SwapRecovery flow handle that instead?

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+    this.logger.info('trying to verify connection to connext');
+    try {
+      await this.sendRequest('/health', 'GET');
+      if (!this.seed) {
```
We could put this check before even pinging the `/health` endpoint, right?

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+      } = transferReceivedRequest;
+      const expectedIncomingTransfer = this.expectedIncomingTransfers.get(rHash);
+      if (!expectedIncomingTransfer) {
+        this.logger.error(`received unexpected incoming transfer with rHash ${rHash}`);
```
Is `error` maybe too severe here? I'm not sure how unexpected it would be to get an incoming transfer we are not expecting.

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+  }
+
+  public removeInvoice = async () => {
+    // not implemented, connext does not use invoices
```
Should we remove from the expected incoming transfers here?

erkarl replied:
Good catch!

sangaman commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+    return {
+      // TODO: verify that it's OK to return
+      // static getTotalTimeLock
+      getTotalTimeLock: () => 101,
```
Does 101 have any significance for Connext?

erkarl replied:
Not really. 101 should be good enough for Connext. I've updated the comment above (not sure why it's not showing up on github).

sangaman reviewed (commented):
Looks really good, I have mostly comments or questions. I am honestly still a bit shaky on the flow regarding incoming transfers and handling preimages, so I'll try to review that part again with fresh eyes after some of these questions are answered.

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

erkarl commented on lib/connextclient/README.md:
````diff
@@ -0,0 +1,71 @@
+# Connext
+
+## Installing Connext node for local development
+```bash
````
For sure. Changed, not sure why it's not showing up as resolved.

Commit:
fixup! feat: add Connext SwapClient

erkarl commented on lib/http/HttpServer.ts:
```diff
@@ -61,6 +61,58 @@ class HttpServer {
+    } catch (err) {
+      statusCode = 400;
+      resJson = { message: JSON.stringify(err), retry: false };
+    }
```
Added a TODO.

Commit:
fixup! feat: add Connext SwapClient

erkarl commented on lib/http/HttpService.ts:
```diff
@@ -19,6 +24,51 @@ class HttpService {
+
+  public providePreimage = async (preimageRequest: ConnextPreimageRequest): Promise<any> => {
+    if (
+      preimageRequest.data && preimageRequest.data.newState
```
Good point. Made the type properties optional to enforce runtime checks - also for the TransferReceivedRequest.

Commit:
fixup! feat: add Connext SwapClient

erkarl commented on lib/http/HttpService.ts:
```diff
@@ -19,6 +24,51 @@ class HttpService {
+      });
+      return { success: true };
+    } else {
+      return { success: false };
```
I removed the success/empty responses all together since the rest api client doesn't process them. All we care about is the response code for now.

erkarl commented on lib/service/Service.ts:
```diff
@@ -258,14 +264,21 @@ class Service {
-        currency,
-        pushAmount,
-      });
+      if (nodeIdentifier === 'Connext') {
```
I'm not convinced this is a better solution. 

Also, a bit foggy on how the implementation details will look like. The peer has to be resolved on the service level since SwapClientManager has no access to Pool (imho, that's a good thing). Let's sync and talk this through offline.

sangaman replied:
Current approach looks great, exactly what I had in mind.

Commit:
fixup! feat: add Connext SwapClient

erkarl commented on lib/Xud.ts:
```diff
@@ -183,6 +184,12 @@ class Xud extends EventEmitter {
 
+      // We initialize Connext separately because it
+      // requires a NodeKey.
+      await this.swapClientManager.initConnext(
```
I originally had this logic in the unlock logic of `InitService`, but then I realized that for the `noencrypt=true` mode that logic is not being used.

erkarl replied:
Created an issue for this: https://github.com/ExchangeUnion/xud/issues/1494 - definitely want this done correctly rather than having a TODO in the code.

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+): Promise<string> => {
+  // create an observable that emits values when a preimage
+  // event is triggered on the client
+  const preimage$: Observable<ProvidePreimageEvent> = fromEvent(client, 'preimage');
```
We could do that, but then we'd have to to create the `preimage$` when starting/initializing the `ConnextClient` and attach it to the class itself:

```
private preimage$: Observable<ProvidePreimageEvent>;
...
this.preimage$ = fromEvent(this, 'preimage');
```
 
We could do that, but what would be the benefit of doing so? Keep in mind that observables by default are cold - meaning that it will not attach event listeners until there's a consumer:

```
observable.subscribe(...)
```

or 
```
observable.toPromise()
```

Commit:
fixup! feat: add Connext SwapClient

erkarl commented on lib/connextclient/ConnextClient.ts:
```diff
@@ -0,0 +1,699 @@
+    });
+  }
+
+  public setSeed = (seed: string) => {
```
Yep, sure sure. That's the ideal way. Added the TODO. I think it's directly related to: https://github.com/ExchangeUnion/xud/issues/1494

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

Commit:
fixup! feat: add Connext SwapClient

raladev commented:
Mention about minor issues: https://github.com/ExchangeUnion/xud/pull/1424#issuecomment-614198780

raladev commented:
Also it seems we need to change swap client for DAI, WETH or change predefined currencies and pairs in xud-docker image.
![Screenshot from 2020-04-23 16-42-24](https://user-images.githubusercontent.com/29906866/80106255-4f3f1b80-8582-11ea-889f-83278145cb3f.png)

kilrau commented:
`getinfo` status things to be fixed:

- [ ] Raiden still showing
```
simnet > getinfo

General XUD Info
┌───────────────┬───────────────────────────────────────────────────────────────────────┐
│ Status        │ Raiden: Could not fetch status informatio                             │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
```

- [ ] connext statically shows 1 channel even though there definitely is no channel:

```
Connext info:
┌──────────┬────────────────────────────────────────────┐
│ Status   │ Ready                                      │
├──────────┼────────────────────────────────────────────┤
│ Version  │ 6.0.4                                      │
├──────────┼────────────────────────────────────────────┤
│ Address  │ 0x1dDFAeBFE4673a16411E11020F1332E38360a895 │
├──────────┼────────────────────────────────────────────┤
│ Channels │ Active: 1 | Pending: 0 | Closed: 0         │
├──────────┼────────────────────────────────────────────┤
│ Network  │  simnet                                    │
└──────────┴────────────────────────────────────────────┘ 
```

Commit:
feat(simnet): automatic channels for Connext

This commit adds logic to automatically requests ETH funds from our
simnet wallet and open a channel to Connext node. This logic is only
initialized when xud is running in simnet mode and Connext SwapClient is
configured.

raladev commented:
Cant perform ETH/BTC swap in Alice-Bob schema (without xud1).

Steps:
1. simnet docker installation for Alice + wait LTC/BTC Active channels
2. simnet docker installation for Bob + wait LTC/BTC Active channels
3. Close all channels
4. Get 2.5 ETH through faucet as Alice
5. Open 0.35 BTC channel from Bob to Alice
6. Place sell 0.1 ETH/BTC 0.1 1 order as Alice
7. Fill it as Bob - place buy 0.1 ETH/BTC 0.1 1 order

Actual result:
(xud of Bob said successful swap + getbalance returns updated balances) but swap stucks because of 500 errors for Alices connext client.

Screens:
balances: left - Alice, right - Bob - [20](https://user-images.githubusercontent.com/29906866/80131104-655cd400-85a2-11ea-9f06-a8031615a336.png)
lncli listchannels + Bob has trade, Alice has not - [21](https://user-images.githubusercontent.com/29906866/80131063-5413c780-85a2-11ea-8ea6-092ac876a7b5.png)
connext errors during swap - [22](https://user-images.githubusercontent.com/29906866/80131239-9a692680-85a2-11ea-92bd-3fa27bc1869b.png)

Logs of Alice - [maker_logs_2.log](https://github.com/ExchangeUnion/xud/files/4524245/maker_logs_2.log)
Logs of Bob - <>

erkarl commented:
> `getinfo` status things to be fixed:

@kilrau what does your `xud.conf` look like? Raiden is removed in `docker/connext` branch, but not from xud codebase. In a fresh environment this should not be the case.

Commit:
fixup! feat: add Connext SwapClient

sangaman reviewed (approved):
Code looks great to me, all concerns above are addressed or make sense to handle in separate PRs. I'll keep testing this but the overall approach is good for me.

Commit:
Merge pull request #1503 from ExchangeUnion/connext-with-channels

feat(simnet): automatic channels for Connext

kilrau commented:
> @kilrau what does your `xud.conf` look like? Raiden is removed in `docker/connext` branch, but not from xud codebase. In a fresh environment this should not be the case.

A fresh (just pulled now) xud-docker simnet (`connext-yy`) -> default simnet conf, no raiden settings enabled. Fresh environment, fresh images, it's definitely still showing raiden status under `General XUD Info` in https://github.com/ExchangeUnion/xud-docker/tree/connext-yy:
```
simnet > getinfo

General XUD Info
┌───────────────┬───────────────────────────────────────────────────────────────────────┐
│ Status        │ LND-BTC: lnd-BTC has no active channels                               │
│               │ LND-LTC: lnd-LTC has no active channels                               │
│               │ Raiden: Could not fetch status informatio                             │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Alias         │ TruthMoral                                                            │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Node Key      │ 02bd97b03705901549a92b7665c8bf176846f2eadf87e80605a3c69dadd7ad09b4    │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Address       │ 02bd97b03705901549a92b7665c8bf176846f2eadf87e80605a3c69dadd7ad09b4    │
│               │ @iltkag6zmfjroq4laeal5audq2jyzj3aisqeqwtxsdtx2vsqgj6kbqid.onion:28885 │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Network       │ simnet                                                                │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Version       │ 1.0.0-beta.2-c2b1e39                                                  │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Peers         │ 3                                                                     │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Pairs         │ 5                                                                     │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Own orders    │ 0                                                                     │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Peer orders   │ 0                                                                     │
├───────────────┼───────────────────────────────────────────────────────────────────────┤
│ Pending swaps │ []                                                                    │
└───────────────┴───────────────────────────────────────────────────────────────────────┘ 
```
